client/src/main/scala/com/dwolla/cloudflare/RateLimitClient.scala

```diff
@@ -63,6 +63,17 @@ class RateLimitClientImpl[F[_] : Sync](executor: StreamingCloudflareApiExecutor[
 
   def delete(zoneId: String, rateLimitId: String): Stream[F, String] = Stream.eval(deleteF(zoneId, rateLimitId))
 
+  implicit val rateLimitActionDtoDecoder: Decoder[RateLimitActionDTO] = (c: HCursor) => for {
+    mode ← c.downField("mode").as[String]
+    timeout ← c.downField("timeout").as[Option[Int]]
+    response ← c.downField("response").as[Option[RateLimitActionResponseDTO]]
+  } yield (mode, timeout) match {
+    case ("challenge", None) ⇒ ChallengeRateLimitActionDTO
+    case ("js_challenge", None) ⇒ JsChallengeRateLimitActionDTO
+    case (_, Some(t)) ⇒ TimeoutRateLimitActionDTO(mode, t, response)
+    case (_, None) ⇒ throw RateLimitMissingTimeoutException()
```

Comment:
I don't think we want to throw an exception here, so that this stays a pure function. It'd be better to put the exception in a `Left` and return it. (`Decoder.Result` is a type alias for `Either`, so the success cases would have to be wrapped in `Right`s as well.)

Reply:
Returned a Left for the exception, but am unable to get to that exception in the test, it just returns a org.http4s.InvalidMessageBodyFailure. Look at RateLimitClientSpec for what I'm talking about.